### PR TITLE
Expose OpenSSL interfaces for base64 coding.

### PR DESCRIPTION
**Pull request by adamreichold**
This exposes the `EVP_EncodeBlock` and `EVP_DecodeBlock` API of OpenSSL which can be used for base64 coding.

It seems reasonable to include those as they have no interactions with and hence unintended side-effects on the rest of the library and it would allow crates already depending on this crate to avoid additional dependencies for base64 coding.

### CODE REVIEW

File: openssl/src/base64.rs [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    // leading and trailing whitespace, but never more.
+    // `EVP_DecodeBlock` will only write to not read from `out`.
+    unsafe {
+        let out_len = cvt_n(ffi::EVP_DecodeBlock(out.as_mut_ptr(), src.as_ptr(), src_len))?;
+        out.set_len(out_len.try_into().unwrap());
+        Ok(out)
+    }
+}
+
+fn encoded_len(src_len: c_int) -> Option<usize> {
```
Comment (sfackler) [inline openssl/src/base64.rs]: Why does this take a c_int and return a usize?

Comment (adamreichold) [reply openssl/src/base64.rs]: I need the return value to `reserve`/`resize` the `out` buffer, so I need an `usize`. But the computation must succeed using a `c_int` for the OpenSSL calls to return valid counts otherwise I am not convinced I can rely on those counts for `set_len`/`truncate`.

In an earlier version, I used `c_int -> Option<c_int>` and had `encoded_len(src_len).unwrap().try_into().unwrap()` on the outside which seemed cumbersome but indeed less surprising. I can certainly changed this back if you prefer or I can add comment explaining the odd types better?

Comment (adamreichold) [reply openssl/src/base64.rs]: But I just noticed that the CI builds failed as the use Rust 1.33 which I guess is this libraries current minimum supported Rust version so I guess I have to remove the `TryInto` usage in any case?

Comment (adamreichold) [reply openssl/src/base64.rs]: Pushed a version that builds using Rust 1.33 and added a comment explaining the odd types.

Would you prefer the `usize_to_c_int` and `c_int_to_usize` in the top-level module or replaced by plain asserts as used in the other modules?

File: openssl/src/base64.rs [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    let mut out = Vec::new();
+    out.reserve(len);
+
+    // SAFETY: `encoded_len` ensures space for 4 output characters
+    // for every 3 input bytes including padding and nul terminator.
+    // `EVP_EncodeBlock` will write only single byte ASCII characters.
+    // `EVP_EncodeBlock` will only write to not read from `out`.
+    unsafe {
+        let out_len = ffi::EVP_EncodeBlock(out.as_mut_ptr(), src.as_ptr(), src_len);
+        out.set_len(c_int_to_usize(out_len).unwrap());
```
Comment (sfackler) [inline openssl/src/base64.rs]: This can just be `out_len as usize`.

File: openssl/src/base64.rs (hunk 1 above)
Comment (sfackler) [reply openssl/src/base64.rs]: It seems best to mirror the rest of the codebase.

File: openssl/src/base64.rs (hunk 2 above)
Comment (adamreichold) [reply openssl/src/base64.rs]: Indeed. Fixed this and the checked cast in `{de,en}coded_len` that was actually unnecessary as well as we always call this with non-negative values and the arithmetic is checked.

File: openssl/src/base64.rs (hunk 1 above)
Comment (adamreichold) [reply openssl/src/base64.rs]: Pushed a version that hopefully better mirrors the style of the rest of the code base.

**Comment by adamreichold** [comment]
During the last rebase, I also noticed that I actually forgot the `#[test]` attribute on two of tests which I now added but which also showed that I forgot to strip the zero padding from the decoded output which I added as well.

File: openssl/src/base64.rs [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    // for every 4 input characters including padding.
+    // `EVP_DecodeBlock` can write fewer bytes after stripping
+    // leading and trailing whitespace, but never more.
+    // `EVP_DecodeBlock` will only write to not read from `out`.
+    unsafe {
+        let out_len = cvt_n(ffi::EVP_DecodeBlock(out.as_mut_ptr(), src.as_ptr(), src_len))?;
+        out.set_len(out_len as usize);
+    }
+
+    if src.ends_with("=") {
```
Comment (sfackler) [inline openssl/src/base64.rs]: What is this doing?

Comment (sfackler) [reply openssl/src/base64.rs]: Oh right, null padding.
